cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -1017,69 +1089,67 @@ class AzureFileSystem::Impl {
             exception);
       }
     } else {
-      auto container_client =
-          blob_service_client_->GetBlobContainerClient(location.container);
-      Azure::Storage::Blobs::ListBlobsOptions options;
-      options.Prefix = internal::EnsureTrailingSlash(location.path);
-      // https://learn.microsoft.com/en-us/rest/api/storageservices/blob-batch#remarks
-      //
-      // Only supports up to 256 subrequests in a single batch. The
-      // size of the body for a batch request can't exceed 4 MB.
-      const int32_t kNumMaxRequestsInBatch = 256;
-      options.PageSizeHint = kNumMaxRequestsInBatch;
+      return DeleteDirContentsWihtoutHierarchicalNamespace(location, true);
+    }
+  }
+
+  Status DeleteDirContents(const AzureLocation& location, bool missing_dir_ok) {
+    if (location.container.empty()) {
+      return internal::InvalidDeleteDirContents(location.all);
+    }
+    if (location.path.empty()) {
+      return internal::InvalidDeleteDirContents(location.all);
```

Comment:
Shouldn't this be interpreted as deleting the contents of the container?

Reply:
We have a special API for this case: `DeleteRootDirContents()`

https://github.com/apache/arrow/blob/63353baf1cda1d1fc7bb614ce01558c12990e073/cpp/src/arrow/filesystem/filesystem.h#L233-L249

So I think that we should return an error here.

Reply:
But what is the root dir? Doesn't that mean *all the containers*?

`s3fs` maps the root-dir concept to *all containers*:
https://github.com/apache/arrow/blob/main/cpp/src/arrow/filesystem/s3fs.cc#L2762

Refusing to delete contents of a container can be very limiting as `CreateDir(container/)` will create a container.

Reply:
Can we clarify this discussion?

What should be "root dir"?

1. ` `: Empty location (no container and no path)
2. `container` and `container/`: Container only
3. `container/path`: Container and path

My opinion and the current implementation:

1. Root dir
2. Root dir
3. Not root dir

FYI: s3fs:

1. ` `: Empty path (no bucket and no key): Root dir
2. `bucket` and `bucket/`: Bucket only: Not root dir
3. `bucket/key`: Bucket and key: Not root dir


@Tom-Newton What do you think about this as an Azure Blob Storage user? I'm not an Azure Blob Storage user.

Reply:
I would consider the storage account the root dir. So that means only  `1` is the root dir. I believe this is how the fsspec implementation and the original arrow azurefs PR did it. 

I think when comparing azure to s3, the Azure storage account is analogous to the S3 bucket and the Azure container is an extra level that S3 doesn't have. I think `1` doesn't really reference any storage on S3.

Reply:
OK. I'll do it.

Reply:
> I think when comparing azure to s3, the Azure storage account is analogous to the S3 bucket and the Azure container is an extra level that S3 doesn't have. I think 1 doesn't really reference any storage on S3.

Azure Containers are S3 Buckets. Azure Blob Names are S3 keys. `s3fs` has an option [1] that forbids/allows creation of new buckets for when the user asks for the creation of a `/newbucket` directory [2].

[1] https://github.com/apache/arrow/blob/main/cpp/src/arrow/filesystem/s3fs.cc#L1820
[2] https://github.com/apache/arrow/blob/main/cpp/src/arrow/filesystem/s3fs.cc#L2672

Reply:
There is also an `allow_bucket_deletion` [1] in s3fs. An analogous to that, but for Azure (let's say `allow_container_deletion`) would be relevant in this PR.

[1] https://github.com/apache/arrow/blob/main/cpp/src/arrow/filesystem/s3fs.cc#L2718

Reply:
Probably I need to review the PR properly to understand the context. I don't feel strongly.

Reply:
The `allow_bucket_deletion` option is for `DeleteDir()` not `DeleteDirContents()`.
And this PR is for `DeleteDirContents()`.
So it should be discussed separately.
I've opened a new issue for it: #38999
We should discuss it further on the issue not here.

Reply:
Overall I'm happy with how this is currently implemented. 

For the sake of argument though... I think `allow_container_deletion` is relevant to `DeleteDirContents` in the case of `location.container.empty()`. I would expect the behaviour in this case to be to delete all containers in the storage account. However, I'm happy to stick with just returning an error because it seems quite dangerous.